modules/thread-pings/src/main/java/net/neoforged/camelot/module/threadpings/ThreadPingsCommand.java

```diff
@@ -227,32 +294,89 @@ private static CommonResult executeCommon(SlashCommandEvent event) {
+                threadPingsExemptions -> threadPingsExemptions.query(channelId));
+        final FilteredRoles exemptRoles = filterRoles(rawExemptRoles, jda);
+
+        return new ChannelPingsData(pingRoles.known(), exemptRoles.known(), pingRoles.unknown(), exemptRoles.unknown());
+    }
+
+    record ChannelPingsData(List<Role> pingRoles, List<Role> exemptRoles, List<Long> unknownPingRoles,
+                            List<Long> unknownExemptRoles) {
+    }
+
+    private static FilteredRoles filterRoles(List<Long> rawRoles, JDA jda) {
+        final var known = new ArrayList<Role>();
+        final var unknown = new ArrayList<Long>();
+        for (long rawRoleId : rawRoles) {
+            final @Nullable Role roleById = jda.getRoleById(rawRoleId);
+            if (roleById != null) {
+                known.add(roleById);
+            } else {
+                unknown.add(rawRoleId);
+            }
+        }
+        return new FilteredRoles(known, unknown);
     }
 
-    record CommonResult(SlashCommandInteraction interaction, GuildChannelUnion channel, List<Role> roles) {
+    record FilteredRoles(List<Role> known, List<Long> unknown) {
     }
 
-    private static String buildMessage(String underText, List<? extends IMentionable> roles) {
+    private static String buildMessage(String place, List<? extends IMentionable> roles, @Nullable List<? extends IMentionable> exemptRoles) {
         final StringBuilder builder = new StringBuilder();
 
-        builder.append("The following roles will be mentioned in public threads created under ")
-                .append(underText)
-                .append(":\n");
+        builder.append("### For __")
+                .append(place)
+                .append("__:")
+                .append('\n');
+
+        builder.append("The following roles will be mentioned in newly-created public threads:")
+                .append('\n');
 
         if (roles.isEmpty()) {
             builder.append("_None._");
         } else {
             builder.append(roles.stream().map(IMentionable::getAsMention).collect(Collectors.joining(", ")));
         }
+        builder.append('\n');
+
+        if (exemptRoles != null) {
+            builder.append('\n')
+                    .append("The following roles are *exempt* from being mentioned in newly-created public threads:")
+                    .append('\n');
+
+            if (exemptRoles.isEmpty()) {
+                builder.append("_None._");
+            } else {
+                builder.append(exemptRoles.stream().map(IMentionable::getAsMention).collect(Collectors.joining(", ")));
+            }
+        }
 
         return builder.toString();
     }
+
+    private static void applyNewRoles(long channelId, List<Long> roleIds, LongFunction<List<Long>> query,
+                                      BiConsumer<Long, Long> adder, BiConsumer<Long, Long> remover) {
```

Comment:
Created a specialized FI type for these biconsumers as which one is the role and which is the channel is confusing with the generic type.